### ISSUE TITLE
Memory leak with video

### ISSUE THREAD

**Issue by wolfgang371** [allo-/virtual_webcam_background#71]
When run with an .mp4 video in the "empty" layer I quickly run out of memory. With a still picture it's fine.

```
video
	  PID  VDATA VSTACK  VSIZE  RSIZE  PSIZE  VGROW  RGROW SWAPSZ  MEM CMD
14:22:	21271   2.6G   140K   9.8G 893.3M     0K 100.5M 100.2M     0K   6% python
14:24:	21271   3.8G   140K  11.0G   2.1G     0K 19740K 11112K     0K  13% python
14:29:	21271   6.5G   140K  13.7G   4.8G     0K 163.0M 186.3M     0K  31% python
14:32:	21271   8.0G   140K  15.1G   6.2G     0K 76444K 75188K     0K  40% python

image
14:32:	21984   2.2G   136K   8.4G 592.6M     0K   8.4G 592.6M     0K   4% python
14:49:	21984   2.2G   136K   8.4G 607.9M     0K -84.4M -81.1M     0K   4% python
```

**Comment by allo-** [comment]
Please post your config.

- Does it still increase after the video was played one time or does it only increase until the full video is loaded?
- Does it increase when the video is reloaded (after changing the config) or does it even decrease?

The video is stored frame by frame (optionally skipping frames) in an array, which is inefficient for longer videos, but should not increase after the full video is stored. Reloading the config resets the list and the python garbage collection should free the memory, but the program will store the frames in a new list again.

**Comment by wolfgang371** [comment]
Hello allo-, here is my config:
```
segmentation_threshold: 0.65
blur: 20
erode: 10
dilate: 15
#width: 1280
#height: 720
#model: mobilenet
internal_resolution: 0.7
virtual_video_device: "/dev/video0"
real_video_device: "/dev/video1"
average_masks: 3
#mjpeg: False
layers:
  #- input: [["blur", 10]]
  #- "empty": [["image", "images/somepic.jpg"]] #, ["roll", 5, 0]]
  - "empty": [["video", "images/vid_60min.mp4"]]
  #- "empty": [["video", "images/vid_6min.mp4"]]
  - "foreground": [["blur", 1]]
```

So the initial video is really 60 minutes long.
I did some tests wrt. to your questions:
- saving & reloading config: does not decrease the memory size
- switching to image after a couple of minutes of video, saving & reloading: does not decrease the memory size
- using the 6 minute video, after looping: does not decrease the memory size, but seems to stay at the maximum size after 6 min.

So in conclusion the video length plays a crucial role and the GC doesn't clean..

**Comment by allo-** [comment]
> So the initial video is really 60 minutes long.

This is the problem. The default is "lazy loading", but this only means the video is decoded as needed and not that old frames are discarded. So After 60 minutes, the memory consumption should stop increasing, but by then, the script probably has used a lot of memory to store every frame of the video.

The video filter is adapted from the image filter that allows loading an animation from a set of images, so it is not optimized for longer videos at all.

An implementation for longer videos would need to use the generator without filling an array that is used as a ring buffer and re-open the video file when the last frame is read.

Instead of appending the image to the list here
https://github.com/allo-/virtual_webcam_background/blob/eb9c6ed14075f7f2897bfdc38597018b886e58d7/filters/video.py#L109

And accessing it by index here
https://github.com/allo-/virtual_webcam_background/blob/eb9c6ed14075f7f2897bfdc38597018b886e58d7/filters/video.py#L113

One would just store the image in the `frame` variable and create a new `lazy_load_video` generator when the StopIteration exception is raised.

**Comment by allo-** [comment]
Please test if https://github.com/allo-/virtual_webcam_background/tree/lazy_video works for you.
This branch only stores the current frame by default.

The change could result in a short interruption at the end of the video when the file is opened again, but many videos will not be a seamless loop anyway.

**Comment by wolfgang371** [comment]
thanks, this fixed the issue!
